US#456048 Make IDOL Admin work through Ingress

hasan-mahmood-opentext opened the pull request:
Just a draft as still need to make this configurable and modify the haproxy thing. Had a couple of questions: 

- Do I need to update the idol-library and single-content .tgz files in helm/release? 
- I've tested ingress manually, but is there testing for this and should we add tests?

daniel-lau-mf commented on helm/idol-library/templates/_ingress.tpl:
```diff
@@ -27,6 +27,19 @@ metadata:
       "nginx.ingress.kubernetes.io/rewrite-target" "/$1"
       "nginx.ingress.kubernetes.io/backend-protocol" (ternary "HTTPS" "HTTP" $component.usingTLS)
   ) -}}
+  {{- if and $ingress.exposeIndexPort $ingress.exposeServicePort }}
+    {{- $_ := set $annotations "nginx.ingress.kubernetes.io/configuration-snippet" `more_set_headers "AciPortPath: /content/";
+more_set_headers "IndexPortPath: /index/";
+more_set_headers "ServicePortPath: /content/service/";` -}}
+  {{- else if $ingress.exposeIndexPort }}
+    {{- $_ := set $annotations "nginx.ingress.kubernetes.io/configuration-snippet" `more_set_headers "AciPortPath: /content/";
+more_set_headers "IndexPortPath: /index/";` -}}
+  {{- else if $ingress.exposeServicePort }}
+    {{- $_ := set $annotations "nginx.ingress.kubernetes.io/configuration-snippet" `more_set_headers "AciPortPath: /content/";
+more_set_headers "ServicePortPath: /content/service/";` -}}
+  {{- else }}
+    {{- $_ := set $annotations "nginx.ingress.kubernetes.io/configuration-snippet" `more_set_headers "AciPortPath: /content/";` -}}
+  {{- end -}}
```
This should use the provided values for `ingress.path`, `ingress.indexPath` and `ingress.servicePath`.

Could probably tidy up the logic by using another helm variable to build the snippet value:
```
{{- $headers := print 
    (ternary "" (printf "more_set_headers \"AciPortPath: %s\";\n" $ingress.path)  (empty $ingress.path))
    (ternary "" (printf "more_set_headers \"ServicePortPath: %s\";\n" $ingress.servicePath)  (empty $ingress.servicePath))
    (ternary "" (printf "more_set_headers \"IndexPortPath: %s\";\n" $ingress.indexPath)  (empty $ingress.indexPath))
-}}
```

We don't need the `.exposeBlah` variables if we just check if the supplied path is empty

daniel-lau-mf commented on helm/single-content/values.yaml:
```diff
@@ -125,6 +126,13 @@ ingress:
 
   # -- Ingress controller path for index connections.
   indexPath: "/index/"
+  # -- Whether to expose the index connection through ingress
+  exposeIndexPort: true
+
+ # -- Ingress controller path for service connections.
+  servicePath: "/content/service/"
+  # -- Whether to expose the service connection through ingress
+  exposeServicePort: true
```
Don't think we need the expose parameters - just make the schema accept empty values and test accordingly before creating paths in ingress
```suggestion
 # -- Ingress controller path for service connections. Empty string to disable.
```

daniel-lau-mf commented on helm/single-content/values.schema.json:
```diff
@@ -230,11 +230,23 @@
+                        "servicePath": {
+                            "$ref": "#/$defs/required_string"
+                        },
+                        "exposeIndexPort": {
+                            "type": "boolean"
+                        },
+                        "exposeServicePort": {
+                            "type": "boolean"
```
See comments about removing these

daniel-lau-mf commented:
* Need to update the idol-library version in order to actually publish the ingress template changes (and that means then updating the charts that use it).

daniel-lau-mf commented:
We might need to rethink how the idol-library ingress templating is handled if we want to avoid having to individually update every components ingress template

daniel-lau-mf reviewed (changes requested):
See comments

hasan-mahmood-opentext commented:
Haven't yet addressed your comment https://github.com/opentext-idol/idol-containers-toolkit/pull/48#issuecomment-2061063788, but have addressed everything else. 
I haven't actually tested this yet since making the changes since I think my ingress might be broken again, will try recreating the cluster.

hasan-mahmood-opentext commented:
Okay, I've managed to test and it seems fine, idol-library and single-content updates are done, just need to update the ingress templates for every other chart.

hasan-mahmood-opentext commented:
Sorry, a lot of files to look through, but it's mostly version updates

daniel-lau-mf commented on helm/idol-library/templates/_ingress.tpl:
```diff
@@ -50,6 +56,24 @@ spec:
   rules:
   - http:
+      paths:
+{{- range $portType, $pathKey := (dict 
+    "aci-port" $ingress.path
+    "service-port" $ingress.servicePath
+    "index-port" $ingress.indexPath
+    "metrics-port" $ingress.metricsPath
+    "prometheus-port" $ingress.prometheusPath
+    ) -}}
```
And/Or we take a `paths` argument and add all of those too - I think might need to do that for the distributed-idol exposedContents ingress. Try templating that at the moment and I suspect you won't get the ingress specification you were intending.

daniel-lau-mf commented on helm/distributed-idol/templates/dih/ingress.yaml:
```diff
@@ -21,26 +21,11 @@ spec:
-            name: {{ $component.name }}
-            port:
-              name: aci-port
       - path: {{ include "idol-library.ingress.path" (dict "ingress" $component.ingress "path" "/index/") }}
         pathType: {{ $pathType }}
         backend:
           service:
             name: {{ $root.Values.indexserviceName }}
             port: 
               name: index-port
```
think this was meant to be deleted too?

hasan-mahmood-opentext replied:
ah I kept that because I thought `$root.Values.indexserviceName` was different to `$component.name`?

daniel-lau-mf commented on helm/distributed-idol/values.schema.json:
```diff
@@ -306,7 +317,23 @@
+                        "path": {
+                            "$ref": "#/$defs/required_string"
+                        },
+                        "metricsPaths": {
```
```suggestion
                        "metricsPath": {
```

hasan-mahmood-opentext commented:
I think everything is working as it should now, templating seems to be all good. The only chart that I haven't changed is nifi, but can look into that?

daniel-lau-mf commented on helm/idol-library/templates/_ingress.tpl:
```diff
@@ -50,9 +61,24 @@ spec:
   - http:
+      paths:
+{{- range $portType, $pathKey := $portmapping -}}
+{{- if $pathKey }}
```
metrics/prometheus should be handled by `$paths` pass in mechanism, not by default handling.

hasan-mahmood-opentext replied:
They're handled by the `$portmapping` variable, they're not in the default handling.

daniel-lau-mf replied:
Sorry, had out of date view.

daniel-lau-mf commented on helm/idol-library-example-test/Chart.yaml:
```diff
@@ -11,12 +11,12 @@
 apiVersion: v2
 name: idol-aci-test
-version: 0.1.0
+version: 0.2.0
 appVersion: "24.1"
```
Updating this version doesn't matter too much as this isn't a published chart

hasan-mahmood-opentext replied:
Should I change it back?

daniel-lau-mf commented on helm/idol-library-example-test/values.yaml:
```diff
@@ -99,6 +99,9 @@ ingress:
   # For an OpenShift environment this is required (see https://docs.openshift.com/container-platform/4.11/networking/routes/route-configuration.html#nw-ingress-creating-a-route-via-an-ingress_route-configuration)
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/acitest/"
```
Don't see this in README.md - maybe need to rebuild that?

daniel-lau-mf replied:
Should add dummy indexPath and servicePath in here too

daniel-lau-mf commented on helm/single-content/Chart.yaml:
```diff
@@ -10,12 +10,12 @@
-  version: "0.10.0"
+  version: "0.11.0"
   repository: https://raw.githubusercontent.com/opentext-idol/idol-containers-toolkit/main/helm
 - name: idol-licenseserver
   condition: idol-licenseserver.enabled
```
You'll need to update the idol-licenseserver version dependency too.

daniel-lau-mf commented on helm/idol-library-example-test/values.schema.json:
```diff
@@ -211,8 +211,12 @@
                 "ingress": {
                     "type":"object",
                     "properties":{
+                        "path": {
+                            "$ref": "#/$defs/required_string"
+                        }
                     },
                     "required":[
+                        "path"
                     ],
```
`path` should go in `#/$defs/ingress/Common`, as should `servicePath`. Could put `indexPath` here as it's not valid for all aci servers

daniel-lau-mf replied:
And what gets done here will need reflecting in the other charts

daniel-lau-mf commented on helm/idol-view/values.schema.json:
```diff
@@ -184,7 +184,19 @@
                 },
                 "ingress": {
-                    "$ref": "#/$defs/ingressCommon"
+                    "type":"object",
```
See idol-library-example-test schema comments

daniel-lau-mf commented on helm/idol-view/values.yaml:
```diff
@@ -106,6 +106,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/view/"
```
Should add default servicePath value.

daniel-lau-mf commented on helm/idol-qms/values.yaml:
```diff
@@ -108,6 +108,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/qms/"
```
Should add default servicePath value

daniel-lau-mf commented on helm/idol-omnigroupserver/Chart.yaml:
```diff
@@ -10,12 +10,12 @@
 dependencies:
 - name: idol-library
-  version: "0.10.0"
+  version: "0.11.0"
```
Will need updated idol-licenseserver version too

daniel-lau-mf commented on helm/idol-omnigroupserver/values.yaml:
```diff
@@ -100,6 +100,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/omnigroupserver/"
```
```suggestion
  path: "/omnigroupserver/"

  #-- Ingress controller path for service connections.
  servicePath: "/omnigroupserver/service/"
```

daniel-lau-mf commented on helm/idol-mediaserver/values.yaml:
```diff
@@ -123,6 +123,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/mediaserver/"
```
```suggestion
  path: "/mediaserver/"

  # -- Ingress controller path for service connections.
  path: "/mediaserver/service/"
```

daniel-lau-mf commented on helm/idol-mediaserver/values.schema.json:
```diff
@@ -131,6 +135,9 @@
                 "host": {
                     "type": "string"
                 },
+                "path": {
```
need servicePath here too.
Could possibly rewrite this entire schema to be based on idol-library-example-test

daniel-lau-mf commented on helm/idol-eductionserver/values.yaml:
```diff
@@ -67,6 +72,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/eductionserver/"
```
```suggestion
  path: "/eductionserver/"

  # -- Ingress controller path for service connections.
  servicePath: "/eductionserver/service/"
```

daniel-lau-mf commented on helm/idol-eductionserver/values.schema.json:
```diff
@@ -96,11 +100,15 @@
                 "host": {
                     "type": "string"
+                },
+                "path": {
```
Add servicePath (and/or rewrite/update to idol-library-example-test base schema)

daniel-lau-mf commented on helm/idol-community/Chart.yaml:
```diff
@@ -10,18 +10,18 @@
   repository: https://raw.githubusercontent.com/opentext-idol/idol-containers-toolkit/main/helm
 - name: idol-licenseserver
   condition: idol-licenseserver.enabled
   version: "0.3.0"
```
idol-licenseserver version needs update too

daniel-lau-mf commented on helm/idol-community/values.yaml:
```diff
@@ -107,6 +107,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/community/"
```
```suggestion
  path: "/community/"

  # -- Ingress controller path for service connections.
  servicePath: "/community/service/"
```

daniel-lau-mf commented on helm/idol-community/values.schema.json:
```diff
@@ -230,8 +230,12 @@
                 "ingress": {
                     "type":"object",
                     "properties":{
+                        "path": {
```
See idol-library-example-test schema comments

daniel-lau-mf commented on helm/idol-answerserver/values.yaml:
```diff
@@ -92,6 +92,9 @@ ingress:
   host: ""
 
+  # -- Ingress controller path for ACI connections.
+  path: "/answerserver/"
```
```suggestion
  path: "/answerserver/"

  # -- Ingress controller path for service connections.
  servicePath: "/answerserver/service/"
```

daniel-lau-mf commented on helm/idol-answerserver/values.schema.json:
```diff
@@ -128,6 +132,9 @@
                 "host": {
                     "type": "string"
                 },
+                "path": {
```
add servicePath / update this schema

daniel-lau-mf commented on helm/distributed-idol/values.yaml:
```diff
@@ -133,6 +133,11 @@ content:
+  # -- whether aci/service/index ports are configured to use TLS (https).
+  # If configuring for TLS, then consider setting IDOL_SSL_COMPONENT_CERT_PATH and IDOL_SSL_COMPONENT_KEY_PATH in envConfigMap
+  # to provide required TLS certificates
+  usingTLS: false
```
Needs schema update

daniel-lau-mf commented on helm/idol-licenseserver/Chart.yaml:
```diff
@@ -14,5 +14,5 @@ version: 0.3.0
 description: Creates an IDOL LicenseServer proxy service accessible to other elements of the cluster
```
Need a new release of idol-licenseserver too for the library update

daniel-lau-mf commented on helm/idol-qms/values.schema.json:
```diff
@@ -230,6 +237,9 @@
                 "ingress": {
                     "type":"object",
                     "properties":{
+                        "path": {
+                            "$ref": "#/$defs/required_string"
+                        }
```
Unneeded here as it's in ingressCommon
